server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java

```diff
@@ -0,0 +1,394 @@
+            }
+            if (!uninitializedExtensions.isEmpty()) {
+                logger.info("Loaded all extensions");
+            }
+        } else {
+            logger.info("Extensions.yml file is not present.  No extensions will be loaded.");
+        }
+    }
+    
+    /**
+     * Loads a single extension
+     * @param extension The extension to be loaded
+     */
+    private void loadExtension(Extension extension) throws IOException {
+        try {
+            uninitializedExtensions.add(
+                new DiscoveryExtensionNode(
+                    extension.getName(),
+                    extension.getUniqueId(),
+                    // placeholder for ephemeral id, will change with POC discovery
+                    extension.getUniqueId(),
+                    extension.getHostName(),
+                    extension.getHostAddress(),
+                    new TransportAddress(InetAddress.getByName(extension.getHostAddress()), Integer.parseInt(extension.getPort())),
+                    new HashMap<String, String>(),
+                    Version.fromString(extension.getOpensearchVersion()),
+                    new PluginInfo(
+                        extension.getName(),
+                        extension.getDescription(),
+                        extension.getVersion(),
+                        Version.fromString(extension.getOpensearchVersion()),
+                        extension.getJavaVersion(),
+                        extension.getClassName(),
+                        new ArrayList<String>(),
+                        Boolean.parseBoolean(extension.hasNativeController())
+                    )
+                )
+            );
+            logger.info("Loaded extension: " + extension);
+        } catch (IllegalArgumentException e) {
+            logger.error(e.toString());
+        }
+    }
+
+    public void initialize() {
+        for (DiscoveryNode extensionNode : uninitializedExtensions) {
+            initializeExtension(extensionNode);
+        }
+    }
+
+    private void initializeExtension(DiscoveryNode extensionNode) {
+
+        final TransportResponseHandler<PluginResponse> pluginResponseHandler = new TransportResponseHandler<PluginResponse>() {
+
+            @Override
+            public PluginResponse read(StreamInput in) throws IOException {
+                return new PluginResponse(in);
+            }
+
+            @Override
+            public void handleResponse(PluginResponse response) {
+                for (DiscoveryExtensionNode extension : uninitializedExtensions) {
+                    if (extension.getName().equals(response.getName())) {
+                        extensions.add(extension);
+                        break;
+                    }
+                }
+            }
+
+            @Override
+            public void handleException(TransportException exp) {
+                logger.debug(new ParameterizedMessage("Plugin request failed"), exp);
```

Comment:
Does it mean extension would stay uninitialized? Should we explicitly write the error message into log?

Reply:
The message is written in the `logger.debug`

Reply:
By why debug if this is an error? (at least it look like a failure)

Reply:
Oh so change it to info instead?

Reply:
Yup thats a good point. @ryanbogan lets change it to `logger.error`.